### PR TITLE
feat(cli): improved diagnostics printing

### PR DESCRIPTION
**Pull request by lucacasonato**
This initially uses the new diagnostic printer in `deno lint`,
`deno doc` and `deno publish`. In the limit we should also update
`deno check` to use this printer.

image

### CODE REVIEW

File: cli/tools/lint.rs [hunk 1]
```diff
@@ -110,9 +123,10 @@ pub async fn lint(flags: Flags, lint_flags: LintFlags) -> Result<(), AnyError> {
       let reporter_kind = lint_options.reporter_kind;
       let reporter_lock = Arc::new(Mutex::new(create_reporter(reporter_kind)));
       let lint_rules = get_config_rules_err_empty(lint_options.rules)?;
-      let r = lint_stdin(lint_rules);
-      let success =
-        handle_lint_result(STDIN_FILE_NAME, r, reporter_lock.clone());
+      let file_path = std::env::current_dir().unwrap().join(STDIN_FILE_NAME);
```
Comment (bartlomieju) [inline cli/tools/lint.rs:126]: You could probably get it from `cli_options.initial_cwd` to avoid th unwrap here

File: cli/diagnostics.rs [hunk 2]
```diff
@@ -0,0 +1,641 @@
...
+  let line_end_pos = source.line_end(line_index);
+  let before = source.range_text(&SourceRange::new(line_start_pos, pos));
+  let after = source.range_text(&SourceRange::new(pos, line_end_pos));
+  (before, after)
+}
+
+/// Returns the text of the line that contains the given positions, split at the
+/// given positions.
+///
+/// If the positions are on different lines, this will panic.
```
Comment (bartlomieju) [inline cli/diagnostics.rs:226]: Is this okay? Maybe it should return an option and not panic?

File: cli/diagnostics.rs [hunk 3]
```diff
@@ -0,0 +1,641 @@
...
+impl<T: Diagnostic + ?Sized> Display for DiagnosticDisplay<'_, T> {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    print_diagnostic(f, self.sources, self.diagnostic)
+  }
+}
+
+// error[missing-return-type]: missing explicit return type on public function
+//   at /mnt/artemis/Projects/github.com/denoland/deno/test.ts:1:16
+//    |
+//  1 | export function test() {
```
Comment (bartlomieju) [inline cli/diagnostics.rs:361]: Do you think in the future we could syntax highlight this code?

**Review by bartlomieju — APPROVED** [review]
Looks great, a few nits and questions.

I think we should aim to rewrite TS diagnostics to use this system in the next patch release. Thoughts?

File: cli/diagnostics.rs (hunk 2 above)
Comment (lucacasonato) [reply cli/diagnostics.rs:226]: The caller ensures they are on the same line already.

File: cli/diagnostics.rs (hunk 3 above)
Comment (lucacasonato) [reply cli/diagnostics.rs:361]: Absolutely!

**Pull request merged**